claat/parser/md/parse_test.go

```diff
@@ -0,0 +1,163 @@
+// Copyright 2018 Google Inc. All Rights Reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+
+package md
+
+import (
+	"fmt"
+	"log"
+	"reflect"
+	"strings"
+	"testing"
+	"time"
+
+	"github.com/googlecodelabs/tools/claat/types"
+)
+
+const stdMeta = `---
+id: codelab
+summary: summary
+
+---`
+
+const stdHeader = stdMeta + `
+# Codelab Title
+`
+
+func mustParseCodelab(markup string) *types.Codelab {
+	c, err := parseCodelab(markup)
+	if err != nil {
+		log.Fatalf("Error parsing markup %v: %v", markup, err)
+	}
+
+	return c
+}
+
+func parseCodelab(markup string) (*types.Codelab, error) {
+	r := strings.NewReader(markup)
+	p := &Parser{}
+
+	return p.Parse(r)
+}
+
+func TestHandleCodelabTitle(t *testing.T) {
+	// Set up.
+	title := "Egret"
+	c := mustParseCodelab(fmt.Sprintf("# %s", title))
+
+	if c.Title != title {
+		t.Errorf("[%q] got %v, want %v", title, c.Title, title)
+	}
+}
+
+// The parser assumes
+//   * Any single number is minutes
+//   * xx::yy is mm::ss
+//   * Hours only appear if you have three parts
+func TestProcessDuration(t *testing.T) {
+	tmp := stdHeader + `
+## Step Title
+Duration: %v`
+
+	tests := []struct {
+		in  string
+		out time.Duration
+	}{
+		// Test an easy case.
+		{"1:00", 1 * time.Minute},
+		// Test a weird number case.
+		{"13:37", 14 * time.Minute},
+		// Test that seconds can be longer than a minute.
+		{"00:90", 2 * time.Minute},
+		// Test that minutes can be longer than an hour.
+		{"00:90:00", 90 * time.Minute},
+		// Test zero.
+		{"0", 0},
+		// Test more than two digits in the hours place.
+		{"100:00:00", 100 * time.Hour},
+		// Test an empty string.
+		{"", 0},
+		// Test complete nonsense.
+		{"Complete nonsense.", 0},
+	}
+
+	for i, tc := range tests {
+		content := fmt.Sprintf(tmp, tc.in)
+		c := mustParseCodelab(content)
+		out := time.Duration(c.Duration) * time.Minute
```

Comment:
Damn, I didn't realize `c.Duration` is not `time.Duration`. We should fix it.
Created https://github.com/googlecodelabs/tools/issues/59.